### ISSUE TITLE
Okta access token is undefined on first login after the okta cache is cleared

### ISSUE THREAD

**Issue by SpaceWarf** [okta/okta-oidc-js#679]
## I'm submitting this issue for the package(s):

- [ ] jwt-verifier
- [ ] okta-angular
- [ ] oidc-middleware
- [X] okta-react
- [ ] okta-react-native
- [ ] okta-vue

## I'm submitting a:

- [X] Bug report
- [ ] Feature request
- [ ] Other (Describe below)

## Current behavior

When using a custom login prompt in our application, the first login after a cache wipe encounters (what i assume is) timing issues since the upgrade to async token retrieval in version 1.0.3 (v1.0.2 works as expected). The issue is only present when the okta cache is wiped from the browser's local storage (which is easily reproduceable by opening a new incognito window). I know the login flow actually works because when I get redirected to the login page for the second time if I manually change the url, I have access to all the routes on the website. The problem is only during the short initial redirect period where the access key is undefined.

Here is our login function:
```
    loginSubmit = e => {
        e.preventDefault();
        this.cleanup();

        const { email, password } = this.state;
        const {  auth } = this.props;

        this.setState({ loading: true }, () => {
            oktaAuth.signIn({
                username: email,
                password
            })
                .then(({ sessionToken }) => {
                    ReactGA.event({
                        category: 'User Action',
                        action: 'Login',
                        label: 'User Login'
                    });
                    auth.redirect({ sessionToken });
                })
                .then(this.cleanup)
                .catch(({ errorSummary, name }) => this.setState({ globalError: errorSummary || name, loading: false }));
        });
    };

```
The problem arises when we try to fetch the access token after the redirect to authorize our requests because it returns `undefined`:
```
const getAccessToken = () => _OktaAuth.tokenManager.get('accessToken');
```

## Expected behavior

The access token should be retrievable on redirect without the need for the cache to be present.

## Extra information about the use case/user story you are trying to implement

Here's a diff of the Auth.js file between version 1.0.2 and 1.0.3
![image](https://user-images.githubusercontent.com/10742962/75281004-770d3f80-57dc-11ea-8df4-dd94ca55255b.png)

## Environment

- Package Version: "@okta/okta-react": "1.4.0"
- Browser: Chrome 80.0.3987.122
- OS: MacOS Catalina 10.15.3
- Node version (`node -v`): v12.14.0
- Other:

**Comment by swiftone** [comment]
Internal ref: OKTA-280936

**Comment by aarongranick-okta** [comment]
@SpaceWarf The `TokenManager.get()` method is async, it returns a promise. The tokens are placed into the TokenManager by the `handleAuthentication()` method which is normally called on the callback route. After the callback route logic is complete, there will be another redirect back to the original path, saved with a call to `setFromUri` (if logging in due to hitting a <SecureRoute>) or back to the app root.

In previous versions of `okta-react`, we advised to use a `checkAuthentication()` method within your component's `componentDidUpdate()` method to check authentication state after each render. Since TokenManager is async, it was not possible to know if a user was authenticated on the initial render and every component would render in a loading/indeterminate state until `checkAuthentication()` could set a true/false state on the component.

We have recently released a new version of `okta-react` which overcomes these shortcomings and gives you a much clearer signal of the users auth state. Now your components can render based on a simple state object passed to your component's props via a HOC or an effect hook, with no async methods or component state required.

**Comment by SpaceWarf** [comment]
I'm in the process of upgrading to okta-react 2.0.0 and noticed that when authState in LoginCallback throws, it tries to print an error object inside a `p` tag which throws a React child error.

![image](https://user-images.githubusercontent.com/10742962/76112256-dbd05300-5faf-11ea-9d45-3608fa19a102.png)

line 23-25 of LoginCallback.js
```
  if(authState.error) { 
    return <p>${authState.error}</p>;
  }
```

**Comment by SpaceWarf** [comment]
Now I'm stuck on an AuthSdkError. Once the authService.redirect() is called, the app crashes because the redirect cookie cannot be retrieved.
```
AuthSdkError: Unable to retrieve OAuth redirect params cookie
```
I inspected the cookies in the Chrome devtools and the required cookie is created, but all the " (double quotation marks) characters are replaced with %22, which causes a fatal error when reading the cookie.

**Comment by swiftone** [comment]
@SpaceWarf - There may be some version conflicts.  Can you remove your `node_modules` and any lock files (`yarn.lock` or `package-lock.json`) and reinstall, and see if the same problem occurs?

**Comment by SpaceWarf** [comment]
A clean install of my `node_modules` and `yarn.lock` did not fix anything, i'm still seeing the cookie error. Is this still a valid way to handle authentication? (this code snippet is in a component wrapped with `withOktaAuth`)

```
_OktaAuth.signIn({
    username: email,
    password
})
    .then(({ sessionToken }) => {
        ReactGA.event({
            category: 'User Action',
            action: 'Login',
            label: 'User Login'
        });
        authService.redirect({ sessionToken });
    })
```
Here is our updated `Security` route:
```
<Security
    {...oktaConfig}
    onAuthRequired={onAuthRequired}
>
    <Route exact path="/implicit/callback" component={LoginCallback} />
    <Route path="/" component={App} />
</Security>
```

**Comment by aarongranick-okta** [comment]
@SpaceWarf that login flow is supported, in fact our E2E test harness app exercises this functionality: https://github.com/okta/okta-oidc-js/blob/master/packages/okta-react/test/e2e/harness/src/SessionTokenLogin.js

Everything seems to be correct in your code from what I can see. Perhaps there is a version mismatch with `okta-auth-js` ? Are you creating an instance of OktaAuth from a separate module installation? The React SDK exposes its internal instance as `authService._oktaAuth`, maybe using that instance will help? `authService._oktaAuth.signIn(...`

We will try to reproduce the error on our side. Do you receive the error in Firefox? or only in Chrome?

**Comment by SpaceWarf** [comment]
My current version of `okta-auth-js` is `3.0.0`. Using the exposed instance of OktaAuth did not change anything. The behaviour i'm seeing is that once I login, I correctly get redirected to `/implicit/callback` with the url parameters, but it then redirects me to the login page again. Then, once i'm at the login page again, without doing anything, it redirects me to `/implicit/callback` again and then to the login page again, but with the `AuthSdkError`.

`Login+login action -> LoginCallback -> Login -> LoginCallback -> Login+AuthSdkError`

**Comment by aarongranick-okta** [comment]
@SpaceWarf 
`okta-react` has not migrated to `okta-auth-js@3.0` yet, but this is probably not be the source of the issue.

Is it possible you have 2 copies of the `LoginCallback` component configured in your router? We recommend using the [<Switch>](https://reacttraining.com/react-router/web/api/Switch) component to ensure only one copy renders. A 2nd instance may be in a race/contention with the first.

Also, are you using version 2.0 or 1.4 of `okta-react`? We did fix some issues with this in 2.0, it should definitely not redirect from the callback unless you are considered authenticated. After the callback you are returned to the original route, or base route. It is assumed the logic there will agree that you are authenticated. If for some reason that logic decides you are not authenticated this could begin a new login flow.

If you are able to place a debugger breakpoint inside `okta-auth-js/lib/token.js` on the `getWithRedirect` method you may be able to see why it is beginning a new login flow after redirect. I recommend webpacking your app with the [source-map-loader](https://www.npmjs.com/package/source-map-loader) webpack plugin so you can debug through the original sources.

**Comment by SpaceWarf** [comment]
Yep, there it is. Wrapping the routes in our `<Security>` component within a `<Switch>` completely fixed every issue we had.

```
<Security
    {...oktaConfig}
    onAuthRequired={onAuthRequired}
>
    <Switch>
        <Route exact path="/implicit/callback" component={LoginCallback} />
        <Route path="/" component={App} />
    </Switch>
</Security>
```
For anyone else stumbling here my current setup is:
* `@okta/okta-react v2.0.0`
* `@okta/okta-auth-js v3.0.0`
* Using our own instance of `OktaAuth`, not the one exposed by `authService`

Thanks for the help !

**Comment by aarongranick-okta** [comment]
@SpaceWarf I'm glad you were able to fix your issues. Just want to follow up so we can better understand the issue.
- Will it also work to remove the `<Switch` and instead put `exact` on the `App` route, `/`? We think the issue may have been that the App component and the LoginCallback were both running on the callback route. Using `<Switch` will prevent this, but putting `exact` on the `/` route should also accomplish this.
- Is it necessary to use a separate instance of OktaAuth?  If you have recently wiped nod_modules and reinstalled, the internal okta-auth instance use by okta-react should be at version 2.13.2. there should not be any issues with the redirect cookie in this version. It would be ideal if the fix for your situation did not require running 2 different version of okta-auth-js.

**Comment by SpaceWarf** [comment]
Using the `exact` parameter on the `App` route does not work our setup. The `App` component itself contains a `<Switch>` with more routes and the `Main` component. The `Main` components then contains even more routes and handles a redirect from the `/` route to the user's last active route.

Switch in App.jsx
```
<Switch>
    <Route path="/auth" component={Auth} />
    // omitted routes
    <SecureRoute path="/" component={Main} />
</Switch>
```

Switch in Main.jsx
```
<Switch>
    <Redirect exact from="/" to={`/${lastActiveSection}`} />
    // omitted routes
</Switch>
```

At the end of the day, the `/` path does have the `exact` attribute, it's just not at the first level in the `Security` tag. I honestly could not explain to you why adding the `<Switch>` in the `Security` tag worked, but without it and even with the `exact` parameter on the first level `/` route, it doesn't work at all.

As for the version of `okta-auth-js`, whether we set it to `3.0.0` or `2.13.2` in our `package.json`, it doesn't affect the workflow in any way that I noticed. What I meant was simply that it's easier for me to create my own OktaAuth object in a utility file so I can use specific methods to generate request headers in other parts of the app.

Snippet of our Okta utility file
```
const _OktaAuth = new OktaAuth({ ...oktaConfig });

const getAccessToken = () => _OktaAuth.tokenManager.get('accessToken');

const clearTokens = () => _OktaAuth.tokenManager.clear();

const getAuthorizationHeader = () => getAccessToken()
    .then(token => { // omitted manipulations })
```

Then in our signin workflow I simply import this object from the utility file instead of using the exposed _oktaAuth object from authService.
